File: dotnet/src/HybridRow/Layouts/TypeArgument.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Layouts
{
    using System;
    using System.Diagnostics;
    using Microsoft.Azure.Cosmos.Core;

    [DebuggerDisplay("{this.type == null ? null : ToString()}")]
    public readonly struct TypeArgument : IEquatable<TypeArgument>
    {
        private readonly LayoutType type;
        private readonly TypeArgumentList typeArgs;

        /// <summary>Initializes a new instance of the <see cref="TypeArgument" /> struct.</summary>
        /// <param name="type">The type of the constraint.</param>
        public TypeArgument(LayoutType type)
        {
            Contract.Requires(type != null);

            this.type = type;
            this.typeArgs = TypeArgumentList.Empty;
        }

        /// <summary>Initializes a new instance of the <see cref="TypeArgument" /> struct.</summary>
        /// <param name="type">The type of the constraint.</param>
        /// <param name="typeArgs">For generic types the type parameters.</param>
        public TypeArgument(LayoutType type, TypeArgumentList typeArgs)
        {
            Contract.Requires(type != null);

            this.type = type;
            this.typeArgs = typeArgs;
        }

        /// <summary>The physical layout type.</summary>
        public LayoutType Type => this.type;

        /// <summary>If the type argument is itself generic, then its type arguments.</summary>
        public TypeArgumentList TypeArgs => this.typeArgs;

        public static bool operator ==(TypeArgument left, TypeArgument right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(TypeArgument left, TypeArgument right)
        {
            return !left.Equals(right);
        }

        /// <summary>The physical layout type of the field cast to the specified type.</summary>
        [DebuggerHidden]
        public T TypeAs<T>()
            where T : ILayoutType
        {
            return this.type.TypeAs<T>();
        }

        public override string ToString()
        {
            if (this.type == null)
            {
                return string.Empty;
            }

            return this.type.Name + this.typeArgs.ToString();
        }

        public override bool Equals(object obj)
        {
            if (object.ReferenceEquals(null, obj))
            {
                return false;
            }

            if (obj is TypeArgument ota)
            {
                return this.Equals(ota);
            }

            return false;
        }

        public override int GetHashCode()
        {
            unchecked
            {
                return (this.type.GetHashCode() * 397) ^ this.typeArgs.GetHashCode();
            }
        }

        public bool Equals(TypeArgument other)
        {
            return this.type.Equals(other.type) && this.typeArgs.Equals(other.typeArgs);
        }
    }
}

File: jre/src/main/java/com/azure/data/cosmos/serialization/hybridrow/layouts/TypeArgument.java
//------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
//------------------------------------------------------------

package com.azure.data.cosmos.serialization.hybridrow.layouts;

import static com.google.common.base.Preconditions.checkNotNull;

public final class TypeArgument {

    public static final TypeArgument NONE = new TypeArgument();
    private final LayoutType type;
    private final TypeArgumentList typeArgs;

    private TypeArgument() {
        this.type = null;
        this.typeArgs = null;
    }

    /**
     * Initializes a new instance of the {@link TypeArgument} struct.
     *
     * @param type The type of the constraint.
     */
    public TypeArgument(LayoutType type) {
        checkNotNull(type);
        this.type = type;
        this.typeArgs = TypeArgumentList.EMPTY;
    }

    /**
     * Initializes a new instance of the {@link TypeArgument} struct.
     *
     * @param type     The type of the constraint.
     * @param typeArgs For generic types the type parameters.
     */
    public TypeArgument(LayoutType type, TypeArgumentList typeArgs) {
        checkNotNull(type);
        this.type = type;
        this.typeArgs = typeArgs;
    }

    @Override
    public boolean equals(Object other) {

        if (null == other) {
            return false;
        }

        return other instanceof TypeArgument && this.equals((TypeArgument) other);
    }

    @Override
    public int hashCode() {
        return (this.type.hashCode() * 397) ^ this.typeArgs.hashCode();
    }

    public boolean equals(TypeArgument other) {
        return this.type.equals(other.type) && this.typeArgs.equals(other.typeArgs);
    }

    @Override
    public String toString() {
        if (this.type == null) {
            return "";
        }
        return this.type.name() + this.typeArgs.toString();
    }

    /**
     * The physical layout type.
     */
    public LayoutType type() {
        return this.type;
    }

    /**
     * If the type argument is itself generic, then its type arguments.
     */
    public TypeArgumentList typeArgs() {
        return this.typeArgs;
    }
}